packages/media_core/src/cluster/room/datachannel.rs

```diff
@@ -0,0 +1,331 @@
+            self.unsub_channel(endpoint, channel_id);
+        }
+    }
+}
+
+impl<Endpoint: Debug + Hash + Eq + Copy> TaskSwitcherChild<Output<Endpoint>> for RoomChannel<Endpoint> {
+    type Time = ();
+
+    fn pop_output(&mut self, _now: Self::Time) -> Option<Output<Endpoint>> {
+        loop {
+            match self.switcher.current()?.try_into().ok()? {
+                TaskType::Publisher => {
+                    if let Some(out) = self.publisher.pop_output((), &mut self.switcher) {
+                        if let Output::OnResourceEmpty = out {
+                            if self.is_empty() {
+                                return Some(Output::OnResourceEmpty);
+                            }
+                        } else {
+                            return Some(out);
+                        }
+                    }
+                }
+                TaskType::Subscriber => {
+                    if let Some(out) = self.subscriber.pop_output((), &mut self.switcher) {
+                        if let Output::OnResourceEmpty = out {
+                            if self.is_empty() {
+                                return Some(Output::OnResourceEmpty);
+                            }
+                        } else {
+                            return Some(out);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<Endpoint> Drop for RoomChannel<Endpoint> {
+    fn drop(&mut self) {
+        log::info!("[ClusterRoomDataChannel] Drop {}", self.room);
+        assert!(self.channels.is_empty(), "channels should be empty on drop");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use atm0s_sdn::features::pubsub::{self, ChannelControl};
+    use media_server_protocol::{datachannel::DataChannelPacket, endpoint::PeerId};
+    use sans_io_runtime::TaskSwitcherChild;
+
+    use crate::cluster::{
+        id_generator,
+        room::datachannel::{Output, RoomChannel},
+        ClusterEndpointEvent,
+    };
+
+    #[test]
+    fn sub_unsub() {
+        let now = ();
+        let room_id = 1.into();
+        let mut room = RoomChannel::new(room_id);
+        let endpoint1 = 1;
+        let endpoint2 = 2;
+        let endpoint3 = 3;
+        let key = "test";
+        let key2 = "test2";
+
+        // 1 -> test
+        // 2 -> test
+        // 3 -> test2
+        let channel_id = id_generator::gen_data_channel_id(room_id, key.to_string());
+        let channel_id2 = id_generator::gen_data_channel_id(room_id, key2.to_string());
+
+        // First subscriber will start publish and subscribe on pubsub channel
+        room.on_channel_subscribe(endpoint1, key);
+        assert_eq!(room.pop_output(now), Some(Output::Pubsub(pubsub::Control(channel_id, ChannelControl::PubStart))));
+        assert_eq!(room.pop_output(now), Some(Output::Pubsub(pubsub::Control(channel_id, ChannelControl::SubAuto))));
+        assert_eq!(room.pop_output(now), None);
```

Comment:
A channel subscription means that it should be able to send and receive messages, which is different from the pubsub subscription in this context. So when a new channel (application channel) is created on the server, meaning that the server should start publish and subscribe, am I understanding this correctly?

Reply:
@giangndm can you check on this

Reply:
I have 2 reason for separating between pub and sub
- In some applications we only have view, who don't publish data into server, then them only need sub
- Pub will make server need to notify to other server about new data source, then it shoud be minimal.

So, I suggest give the control to client, developer can free to pub or sub a channel.

Reply:
So I should just expose some kind of API for the pub sub functionality instead of abstracting it into a datachannel, so the client side or the sdk side can do the abstraction later?

Reply:
@luongngocminh yes, I think it is better, if we need to combine it, we can do it in SDK